contracts/libraries/ArbitraryMessage.sol

```diff
@@ -0,0 +1,82 @@
+pragma solidity 0.4.24;
+
+import "../interfaces/IBridgeValidators.sol";
+import "./Message.sol";
+
+library ArbitraryMessage {
+    // layout of message :: bytes:
+    // offset  0: 32 bytes :: uint256 - message length
+    // offset 32: 20 bytes :: address - sender address
+    // offset 52: 20 bytes :: address - executor contract
+    // offset 72: 32 bytes :: bytes32 txHash
+    // offset 104: 32 bytes :: uint256 - gasLimit
+    // offset 136: 1 bytes :: bytes1 - dataType
+    // (optional) 137: 32 bytes :: uint256 - gasPrice
+    // (optional) 137: 1 bytes :: bytes1 - gasPriceSpeed
+
+    // bytes 1 to 32 are 0 because message length is stored as little endian.
+    // mload always reads 32 bytes.
+    // so we can and have to start reading recipient at offset 20 instead of 32.
+    // if we were to read at 32 the address would contain part of value and be corrupted.
+    // when reading from offset 20 mload will read 12 zero bytes followed
+    // by the 20 recipient address bytes and correctly convert it into an address.
+    // this saves some storage/gas over the alternative solution
+    // which is padding address to 32 bytes and reading recipient at offset 32.
+    // for more details see discussion in:
+    // https://github.com/paritytech/parity-bridge/issues/61
+
+    function unpackData(bytes _data, bool applyDataOffset)
```

Comment:
To get the `data` parameter from the bridged message in `unpackData` the following line is used to copy values directly from `calldata`:

```solidity
calldatacopy(dataptr, add(add(68, srcdataptr), dataOffset), datasize)
```

While implementing this I found that when `unpackData` was called on home side from the method `function executeAffirmation(bytes message) external` it was working correctly. But when it was being called from foreign side from method `function executeSignatures(uint8[] vs, bytes32[] rs, bytes32[] ss, bytes message)` the `data` parameter from the message was being filled with incorrects values. Also the values assigned to `data` were different depending on the amount of validator signatures I used on every call. 

So that's why I changed the order of the parameters to:
`function executeSignatures(bytes _data, uint8[] vs, bytes32[] rs, bytes32[] ss) external`

After that change, independently of the numbers of validator signatures, the `data` parameter was being filled with the same incorrect values on every call. I found that there was an offset of 9 positions on the values that were being read from `calldata`, which also could be different depending on the `dataType` of the message (if gas prices values or gas price oracle speed were in the message).

I'm not sure exactly how the parameters from the external methods are managed for `calldata` but it seems to be affecting the position of the information inside `calldata` that's why the previous changes (parameter order change and `applyDataOffset` parameter) were introduced to make `unpackData` method work for home and foreign methods.

Reply:
Hmm. Quite interesting behavior. Let's have it as is so far. Probably later we can investigate it deeply.